system-helper/flatpak-system-helper.c

```diff
@@ -141,6 +311,188 @@ no_progress_cb (OstreeAsyncProgress *progress, gpointer user_data)
+  guint uid_pull_count = 0;
+
+  G_LOCK (ongoing_pulls);
+  for (guint i = 0; i < ongoing_pulls->len; i++)
+    {
+      pull = g_ptr_array_index (ongoing_pulls, i);
+      if (pull->uid == uid)
+        uid_pull_count++;
+    }
+  G_UNLOCK (ongoing_pulls);
+
+  return uid_pull_count;
+}
+
+static void
+subprocess_wait_check_cb (GObject      *source_object,
+                          GAsyncResult *result,
+                          gpointer      user_data)
+{
+  GSubprocess *subprocess = (GSubprocess *) source_object;
+  FlatpakSystemHelperOngoingPull *pull = user_data;
+  g_autoptr(GError) error = NULL;
+
+  if (!g_subprocess_wait_check_finish (subprocess, result, &error))
+    {
+      if (!g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED))
+        {
+          g_dbus_method_invocation_return_gerror (pull->invocation, error);
+          system_helper_ongoing_pull_free (pull);
+        }
+    }
+}
+
+static gboolean
+handle_create_pull_cache_dir (FlatpakSystemHelper   *object,
+                              GDBusMethodInvocation *invocation,
+                              const gchar           *arg_installation)
+{
+  GDBusConnection *connection = g_dbus_method_invocation_get_connection (invocation);
+  g_autoptr(FlatpakDir) system = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autofree gchar *bindfs_path = NULL;
+
+  system = dir_get_system (arg_installation, &error);
+  if (system == NULL)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  bindfs_path = g_find_program_in_path ("bindfs");
+  if (bindfs_path == NULL)
+    {
+      g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_NOT_SUPPORTED,
+                                             "Can't find bindfs in the system.");
+      return TRUE;
+    }
+  else
+    {
+      FlatpakSystemHelperOngoingPull *new_pull;
+      g_autoptr(FlatpakSystemHelperOngoingPull) pull_tmp = NULL;
+      g_autofree gchar *user_dir = NULL;
+      g_autofree gchar *root_dir = NULL;
+      g_autofree gchar *flatpak_dir = NULL;
+      g_autofree gchar *arg_user = NULL;
+      guint uid_pull_count = 0;
+      uid_t uid;
+
+      if (!get_connection_uid (invocation, &uid, &error))
+        {
+          g_dbus_method_invocation_return_gerror (invocation, error);
+          return TRUE;
+        }
+
+      uid_pull_count = get_uid_pull_count (uid);
+      if (uid_pull_count > MAX_PULLS_PER_USER)
+        {
+          g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_LIMITS_EXCEEDED,
+                                                 "Maximum number of pulls per user exceeded for UID: %u", uid);
+          return TRUE;
+        }
+
+      // TODO: If there is a already partial pull available, return that.
+
+      flatpak_dir = g_file_get_path (flatpak_dir_get_path (system));
+      root_dir = g_mkdtemp_full (g_build_filename (flatpak_dir, "repo", "tmp", "flatpak-cache-XXXXXX", NULL), 0755);
```

Comment:
You are right about the first question. Same filesystem etc.

It will get cleaned up `min-free-space-*` gating conditions are hit. They will clear out everything in `repo/tmp`. If they are not hit, they will / can be re-used to resume partial pulls. Checkout `check_partial_pull_availability` function.